jax/_src/typing/dtypes.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2023 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from typing import Any, Literal, Protocol, Union
+import numpy as np
+import ml_dtypes
+
+DType = np.dtype
+
+# TODO(jakevdp, froystig): make ExtendedDType a protocol
+ExtendedDType = Any
+
+class SupportsDType(Protocol):
+  @property
+  def dtype(self) -> DType: ...
+
+# DTypeLike is meant to annotate inputs to np.dtype that return
+# a valid JAX dtype. It's different than numpy.typing.DTypeLike
+# because JAX doesn't support objects or structured dtypes.
+# Unlike np.typing.DTypeLike, we exclude None, and instead require
+# explicit annotations when None is acceptable.
+# TODO(jakevdp): consider whether to add ExtendedDtype to the union.
+DTypeLike = Union[
+  str,            # like 'float32', 'int32'
```

Comment:
Would it be useful to make this a union of literal types iyo?

Reply:
Possibly – this is a public API though, so making it more strictly defined will in general lead to downstream failures that will prevent us from landing this.